### PR TITLE
[SPARK-11206] Support SQL UI on the history server

### PR DESCRIPTION
**Pull request by carsonwang**
On the live web UI, there is a SQL tab which provides valuable information for the SQL query. But once the workload is finished, we won't see the SQL tab on the history server. It will be helpful if we support SQL UI on the history server so we can analyze it even after its execution.

To support SQL UI on the history server:
1. I added an `onOtherEvent` method to the `SparkListener` trait and post all SQL related events to the same event bus.
2. Two SQL events `SparkListenerSQLExecutionStart` and `SparkListenerSQLExecutionEnd` are defined in the sql module.
3. The new SQL events are written to event log using Jackson.
4.  A new trait `SparkHistoryListenerFactory` is added to allow the history server to feed events to the SQL history listener. The SQL implementation is loaded at runtime using `java.util.ServiceLoader`.

### CODE REVIEW

File: core/src/main/scala/org/apache/spark/util/JsonProtocol.scala [hunk 1]
```diff
@@ -504,9 +542,28 @@ private[spark] object JsonProtocol {
       case `executorRemoved` => executorRemovedFromJson(json)
       case `logStart` => logStartFromJson(json)
       case `metricsUpdate` => executorMetricsUpdateFromJson(json)
+      case `sqlExecutionStart` => sqlExecutionStartFromJson(json)
+      case `sqlExecutionEnd` => sqlExecutionEndFromJson(json)
```
Comment (steveloughran) [inline core/src/main/scala/org/apache/spark/util/JsonProtocol.scala]: This mach really needs `case other:String => throw new IOException(s"Unknown Event $other") message, so that when new events are added in future, the old replay code will at least display a meaningful message. Maybe that's for another pull req though.

Comment (carsonwang) [inline core/src/main/scala/org/apache/spark/util/JsonProtocol.scala]: Added that. Thanks.

**Comment by vanzin** [comment]
Hi @carsonwang,

This is a really useful addition, but I'm afraid this approach is not really scalable. For example, you have the same problem with the Streaming tab in the UI, and trying to fix that with your approach would just make things worse.

I actually think the underlying code needs some lower level changes before this can be implemented more cleanly. For example, a few things that I thought about in the past:

- `SparkListenerEvent` should not be a sealed trait; that way streaming events and sql events can extend it without having to exist in core
- `SparkListener` should not be the root of the listener hierarchy; `SparkFirehoseListener` should, and `SparkListener` should be a specialization of that trait that provides more type-specific method.
- events should be written to the logs using a library like Jackson instead of having to modify `JsonProtocol` for each new event. The current approach is just not scalable. The main issue here is backwards compatibility, but I think this is a much needed change in this part of the code.

With those in place, the only thing left is a protocol so that the history server can feed events to backend-specific listeners; that could be done by having the history server expose a new trait that is loaded at runtime using something like `java.util.ServiceLoader`; then there could be an implementation of that trait in sql (and another one in streaming) to process events read from the logs and add stuff to the UI.

What do you think? I really would like to avoid moving more stuff to the core; it feels like this would be a pretty good opportunity to work on some of the outstanding issues with this whole part of the code.

**Comment by carsonwang** [comment]
@vanzin  Thanks a lot for the comment. This sounds great and is very helpful. I agree it is not a good idea to move more stuff to the core. The underlying code change you mentioned sounds a lot of works to do, especially when we consider backwards compatibility. Can we apply part of them so that we can support SQL UI on the history server in this PR and also avoid moving the SQL classes to the core? I was thinking adding a `toJson` method to the `SparkListenerEvent`. As you mentioned when it is not a sealed trait, we can define sql events in the sql module. For existing events in the core, we still use the `JsonProtocol` in the `toJson` method to convert the event to json. For sql events, we can define `toJson` method in the sql module. This way we will be able to write the sql events to the log. Then we can use something you mentioned to feed events from the history server to the sql listener.

I know this might still not be very scalable. But we don't make things worse once we are able to avoid moving sql classes to the core and also support SQL UI on the history server. Do you think this is doable for this ticket?

**Comment by vanzin** [comment]
> I was thinking adding a toJson method to the SparkListenerEvent.

I'm not a big fan of allowing events to write custom serialization code. I think a better approach would be to change `JsonProtocol` to use Jackson when the event doesn't match any of the currently known event types. Then later we could add a flag that makes `JsonProtocol` use Jackson for all event types.

Mostly that's for backwards compatibility, so that event that are currently written to the event log don't change format. We can then document that the old format is deprecated and will be switched to the new, Jackson-based one in a future release.

**Comment by carsonwang** [comment]
@vanzin , your suggestions are very helpful. I have updated the code following the ideas. I use Jackson to serialize the SQL events and also introduce a trait to register the events and feed them to the SQL listener. The SQL implementation of the trait is loaded at runtime using `java.util.ServiceLoader`. Can you please help take a look?

**Comment by carsonwang** [comment]
Jenkins, retest this please.

File: core/src/main/scala/org/apache/spark/util/JsonProtocol.scala [hunk 2]
```diff
@@ -52,7 +55,22 @@ import org.apache.spark.storage._
...
 
-  private implicit val format = DefaultFormats
+  // Events defined in other modules are allowed to be registered through implementing the
+  // SparkListenerEventRegister trait. These events are written to Json using Jackson.
+  val eventRegisters: Iterable[SparkListenerEventRegister] = {
+    val loader = Utils.getContextOrSparkClassLoader
+    ServiceLoader.load(classOf[SparkListenerEventRegister], loader).asScala
+  }
+
+  var eventClasses = new ListBuffer[Class[_]]()
```
Comment (vanzin) [inline core/src/main/scala/org/apache/spark/util/JsonProtocol.scala]: This can be a `val`. I still want to check how this stuff is used, though (haven't made it to that part of the patch yet).

    val eventClasses = eventRegisters.flatMap { r => eventRegister.getEventClasses() }.toSeq

File: core/src/main/scala/org/apache/spark/util/JsonProtocol.scala [hunk 3]
```diff
@@ -96,6 +114,7 @@ private[spark] object JsonProtocol {
         executorMetricsUpdateToJson(metricsUpdate)
       case blockUpdated: SparkListenerBlockUpdated =>
         throw new MatchError(blockUpdated)  // TODO(ekl) implement this
+      case _ => parse(write(event))
```
Comment (vanzin) [inline core/src/main/scala/org/apache/spark/util/JsonProtocol.scala]: Hmm... I'm not sure I like this. This is why you need the class list above, when it would be much nicer to not need it. You can avoid the class list by using the Jackson API directly. For example:

    import com.fasterxml.jackson.annotation._
    import com.fasterxml.jackson.databind.ObjectMapper
    
    @JsonTypeInfo(use=JsonTypeInfo.Id.CLASS, include=JsonTypeInfo.As.PROPERTY, property="Event") class A { }
    
    class B(val bar: String, val baz: Option[String]) extends A
    
    object main {
      def main(args: Array[String]): Unit = {
        val mapper = new ObjectMapper()
        mapper.registerModule(com.fasterxml.jackson.module.scala.DefaultScalaModule)
    
        val b = new B("bar", None)
        val out = mapper.writeValueAsString(b)
    
        val read = mapper.readValue(out, classOf[A]).asInstanceOf[B]
        println(s"$read: ${read.bar} :: ${read.baz}")
      }
    }
    
Basically, annotate the root event (`SparkEventListener`) with `JsonTypeInfo`, and use an `ObjectMapper` to generate and parse the JSON.

I tried but couldn't make this work with `DefaultParams`. Setting just `typeHintFieldName` doesn't work, it seems you need `typeHints`, and for that you need a class list. It seems to just ignore the annotations.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SQLExecution.scala [hunk 4]
```diff
@@ -45,25 +46,14 @@ private[sql] object SQLExecution {
...
       val r = try {
         val callSite = Utils.getCallSite()
-        sqlContext.listener.onExecutionStart(
-          executionId,
-          callSite.shortForm,
-          callSite.longForm,
-          queryExecution.toString,
-          SparkPlanGraph(queryExecution.executedPlan),
-          System.currentTimeMillis())
+        sqlContext.sparkContext.listenerBus.post(SparkListenerSQLExecutionStart(
```
Comment (vanzin) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SQLExecution.scala]: I'll summon @marmbrus to see whether the change in semantics here is ok. The events would now be processed asynchronously instead of in the same thread.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SparkPlanInfo.scala [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+ * :: DeveloperApi ::
+ * Stores information about a SQL SparkPlan.
+ */
+@DeveloperApi
+class SparkPlanInfo(
+    val nodeName: String,
+    val simpleString: String,
+    val children: Seq[SparkPlanInfo],
+    val metrics: Seq[SQLMetricInfo]
+) {
```
Comment (vanzin) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkPlanInfo.scala]: nit: brackets aren't necessary.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/ui/SQLListener.scala [hunk 6]
```diff
@@ -193,37 +224,41 @@ private[sql] class SQLListener(conf: SparkConf) extends SparkListener with Loggi
...
-        // onExecutionEnd happens after all "onJobEnd"s
-        // So we should update the execution lists.
-        markExecutionFinished(executionId)
-      } else {
-        // There are some running jobs, onExecutionEnd happens before some "onJobEnd"s.
-        // Then we don't if the execution is successful, so let the last onJobEnd updates the
-        // execution lists.
-      }
+  override def onOtherEvent(event: SparkListenerEvent): Unit = {
+    event match {
```
Comment (vanzin) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/ui/SQLListener.scala]: minor: you can move this to the previous line and avoid one level of indentation.

    ...: Unit = event match {

File: core/src/main/scala/org/apache/spark/ui/SparkUI.scala [hunk 7]
```diff
@@ -150,7 +151,14 @@ private[spark] object SparkUI {
       appName: String,
       basePath: String,
       startTime: Long): SparkUI = {
-    create(None, conf, listenerBus, securityManager, appName, basePath, startTime = startTime)
+    val sparkUI = create(
+      None, conf, listenerBus, securityManager, appName, basePath, startTime = startTime)
+    JsonProtocol.eventRegisters.foreach { eventRegister =>
+      val listener = eventRegister.getListener()
+      listenerBus.addListener(listener)
+      eventRegister.attachUITab(listener, sparkUI)
```
Comment (vanzin) [inline core/src/main/scala/org/apache/spark/ui/SparkUI.scala]: This doesn't look right. Won't this attach, e.g. the SQL tab to every application in the history server, regardless of whether the app used SQL or not?

File: core/src/main/scala/org/apache/spark/util/JsonProtocol.scala (hunk 3 above)
Comment (steveloughran) [inline core/src/main/scala/org/apache/spark/util/JsonProtocol.scala]: if some more generic handling of json marshalling is in there (nice), why is `SparkListenerBlockUpdated` being excluded? It should just be one of the types marshalled

Comment (carsonwang) [inline core/src/main/scala/org/apache/spark/util/JsonProtocol.scala]: I tried using the `JsonTypeInfo` annotation and it worked. The difference is that it need write the full class name now. The event will look like 
```
"Event":"org.apache.spark.sql.execution.ui.SparkListenerSQLExecutionStart"
```
When deserializing the data, we use reflection to find the class so that the class list is not needed.

Comment (carsonwang) [inline core/src/main/scala/org/apache/spark/util/JsonProtocol.scala]: Hi @steveloughran, I think currently the generic handling is only for new events added in the sub modules.

File: core/src/main/scala/org/apache/spark/ui/SparkUI.scala (hunk 7 above)
Comment (carsonwang) [inline core/src/main/scala/org/apache/spark/ui/SparkUI.scala]: Yes. Ok, I changed to only attach the SQL tab when a `SparkListenerSQLExecutionStart` event is received for the first time.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SQLExecution.scala (hunk 4 above)
Comment (marmbrus) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SQLExecution.scala]: /cc @zsxwing @andrewor14

**Comment by carsonwang** [comment]
Jenkins, retest this please.

**Comment by carsonwang** [comment]
retest this please

**Comment by carsonwang** [comment]
retest this please

**Comment by carsonwang** [comment]
Hi @vanzin, I updated the code to address your comments. Thank you!

File: core/src/main/scala/org/apache/spark/scheduler/SparkListener.scala [hunk 8]
```diff
@@ -19,18 +19,22 @@ package org.apache.spark.scheduler
 
 import java.util.Properties
 
+import com.fasterxml.jackson.annotation.JsonTypeInfo
```
Comment (vanzin) [inline core/src/main/scala/org/apache/spark/scheduler/SparkListener.scala]: nit: should go after scala imports.

File: core/src/main/scala/org/apache/spark/scheduler/SparkListener.scala [hunk 9]
```diff
@@ -131,6 +135,17 @@ case class SparkListenerApplicationEnd(time: Long) extends SparkListenerEvent
 private[spark] case class SparkListenerLogStart(sparkVersion: String) extends SparkListenerEvent
 
 /**
+ * Interface for registering listeners defined in other modules like SQL, which are used to
+ * rebuild the history UI.
+ */
+trait SparkListenerRegister {
```
Comment (vanzin) [inline core/src/main/scala/org/apache/spark/scheduler/SparkListener.scala]: I know it's a little annoying to rename things, but I don't like this name very much... how about `SparkHistoryListenerFactory` or something along those lines?

Also, this trait should probably be `private[spark]`.

File: core/src/main/scala/org/apache/spark/scheduler/SparkListener.scala [hunk 10]
```diff
@@ -131,6 +135,17 @@ case class SparkListenerApplicationEnd(time: Long) extends SparkListenerEvent
...
+  /**
+   * Get a listener used when rebuilding the history UI.
+   */
+  def getListener(conf: SparkConf, sparkUI: SparkUI): SparkListener
```
Comment (vanzin) [inline core/src/main/scala/org/apache/spark/scheduler/SparkListener.scala]: How about `createListeners`, and have it return a `Seq`? Just in case.

File: core/src/main/scala/org/apache/spark/ui/SparkUI.scala [hunk 11]
```diff
@@ -127,6 +130,11 @@ private[spark] object SparkUI {
   val DEFAULT_RETAINED_STAGES = 1000
   val DEFAULT_RETAINED_JOBS = 1000
 
+  val listenerRegisters: Iterable[SparkListenerRegister] = {
```
Comment (vanzin) [inline core/src/main/scala/org/apache/spark/ui/SparkUI.scala]: Do you need to keep this in a `val`? Could it be a local variable in `createHistoryUI`?

File: core/src/main/scala/org/apache/spark/util/JsonProtocol.scala [hunk 12]
```diff
@@ -19,7 +19,8 @@ package org.apache.spark.util
 
 import java.util.{Properties, UUID}
 
-import org.apache.spark.scheduler.cluster.ExecutorInfo
+import com.fasterxml.jackson.databind.ObjectMapper
```
Comment (vanzin) [inline core/src/main/scala/org/apache/spark/util/JsonProtocol.scala]: nit: should go after scala imports.

File: core/src/main/scala/org/apache/spark/util/JsonProtocol.scala [hunk 13]
```diff
@@ -54,6 +57,8 @@ private[spark] object JsonProtocol {
 
   private implicit val format = DefaultFormats
 
+  val mapper = new ObjectMapper().registerModule(DefaultScalaModule)
```
Comment (vanzin) [inline core/src/main/scala/org/apache/spark/util/JsonProtocol.scala]: private?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/metric/SQLMetrics.scala [hunk 14]
```diff
@@ -126,31 +144,25 @@ private[sql] object SQLMetrics {
...
     // 100GB (100MB, 1GB, 10GB)
-    createLongMetric(sc, s"$name total (min, med, max)", stringValue, -1L)
+    createLongMetric(sc, s"$name total (min, med, max)", StaticsLongSQLMetricParam)
+  }
+
+  def getMetricParam(metricParamName: String): SQLMetricParam[SQLMetricValue[Any], Any] = {
+    val longSQLMetricParam = Utils.getFormattedClassName(LongSQLMetricParam)
+    val staticsSQLMetricParam = Utils.getFormattedClassName(StaticsLongSQLMetricParam)
+    val metricParam = metricParamName match {
+        case `longSQLMetricParam` => LongSQLMetricParam
```
Comment (vanzin) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/metric/SQLMetrics.scala]: nit: indented too far

File: sql/core/src/main/scala/org/apache/spark/sql/execution/ui/SQLListener.scala [hunk 15]
```diff
@@ -289,6 +311,38 @@ private[sql] class SQLListener(conf: SparkConf) extends SparkListener with Loggi
 
 }
 
+private[spark] class SQLHistoryListener(conf: SparkConf, sparkUI: SparkUI)
+  extends SQLListener(conf) {
+
+  var sqlTabAttached = false
```
Comment (vanzin) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/ui/SQLListener.scala]: private?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/ui/SQLListener.scala [hunk 16]
```diff
@@ -289,6 +311,38 @@ private[sql] class SQLListener(conf: SparkConf) extends SparkListener with Loggi
...
+      taskEnd.stageId,
+      taskEnd.stageAttemptId,
+      taskEnd.taskInfo.accumulables.map { acc =>
+        (acc.id, new LongSQLMetricValue(acc.update.getOrElse("0").toLong))
+      }.toMap,
+      finishTask = true)
+  }
+
+  override def onOtherEvent(event: SparkListenerEvent): Unit = event match {
+    case executionStart: SparkListenerSQLExecutionStart =>
```
Comment (vanzin) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/ui/SQLListener.scala]: `executionStart` is not used, so you can do `case _: Spark... =>`

**Comment by vanzin** [comment]
Looks ok to me, mostly nits remaining. It would be nice if someone could confirm that the change in how SQL events are delivered is fine.

File: core/src/main/scala/org/apache/spark/scheduler/SparkListener.scala (hunk 9 above)
Comment (carsonwang) [inline core/src/main/scala/org/apache/spark/scheduler/SparkListener.scala]: `SparkHistoryListenerFactory` sounds good.

File: core/src/main/scala/org/apache/spark/ui/SparkUI.scala (hunk 11 above)
Comment (carsonwang) [inline core/src/main/scala/org/apache/spark/ui/SparkUI.scala]: Oh yes, this can be a local variable.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/ui/SQLListener.scala [hunk 17]
```diff
@@ -193,38 +214,39 @@ private[sql] class SQLListener(conf: SparkConf) extends SparkListener with Loggi
...
-      if (!executionUIData.hasRunningJobs) {
-        // onExecutionEnd happens after all "onJobEnd"s
-        // So we should update the execution lists.
-        markExecutionFinished(executionId)
-      } else {
-        // There are some running jobs, onExecutionEnd happens before some "onJobEnd"s.
-        // Then we don't if the execution is successful, so let the last onJobEnd updates the
-        // execution lists.
+  override def onOtherEvent(event: SparkListenerEvent): Unit = event match {
+    case executionStart: SparkListenerSQLExecutionStart =>
```
Comment (chenghao-intel) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/ui/SQLListener.scala]: case SparkListenerSQLExecutionStart(executionId, description, ..) =>

File: sql/core/src/main/scala/org/apache/spark/sql/execution/ui/SQLListener.scala [hunk 18]
```diff
@@ -193,38 +214,39 @@ private[sql] class SQLListener(conf: SparkConf) extends SparkListener with Loggi
...
+        executionStart.details,
+        executionStart.physicalPlanDescription,
+        physicalPlanGraph,
+        sqlPlanMetrics.toMap,
+        executionStart.time)
+      synchronized {
+        activeExecutions(executionStart.executionId) = executionUIData
+        _executionIdToData(executionStart.executionId) = executionUIData
+      }
+    case executionEnd: SparkListenerSQLExecutionEnd => synchronized {
```
Comment (chenghao-intel) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/ui/SQLListener.scala]: Same here

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SQLExecution.scala (hunk 4 above)
Comment (chenghao-intel) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SQLExecution.scala]: This looks good to me, as `SQLListener` will handle the event of `SparkListenerSQLExecutionStart` properly.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/metric/SQLMetrics.scala [hunk 19]
```diff
@@ -104,21 +104,39 @@ private class LongSQLMetricParam(val stringValue: Seq[Long] => String, initialVa
   override def zero: LongSQLMetricValue = new LongSQLMetricValue(initialValue)
 }
 
+private[sql] object LongSQLMetricParam extends LongSQLMetricParam(_.sum.toString, 0L)
```
Comment (chenghao-intel) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/metric/SQLMetrics.scala]: Nit: change the object name `LongSQLMetricParam ` to something else? Compare to `StaticsLongSQLMetricParam`.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/metric/SQLMetrics.scala [hunk 20]
```diff
@@ -91,7 +91,7 @@ private[sql] class LongSQLMetric private[metric](name: String, param: LongSQLMet
   }
 }
 
-private class LongSQLMetricParam(val stringValue: Seq[Long] => String, initialValue: Long)
+private[sql] class LongSQLMetricParam(val stringValue: Seq[Long] => String, initialValue: Long)
```
Comment (chenghao-intel) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/metric/SQLMetrics.scala]: Can we keep the original visibility, by removing the `[sql]`?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/ui/SQLListener.scala (hunk 17 above)
Comment (carsonwang) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/ui/SQLListener.scala]: done

File: sql/core/src/main/scala/org/apache/spark/sql/execution/ui/SQLListener.scala (hunk 18 above)
Comment (carsonwang) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/ui/SQLListener.scala]: done

File: sql/core/src/main/scala/org/apache/spark/sql/execution/metric/SQLMetrics.scala (hunk 20 above)
Comment (carsonwang) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/metric/SQLMetrics.scala]: Removed that after removing another two `[sql]`.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/metric/SQLMetrics.scala (hunk 19 above)
Comment (carsonwang) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/metric/SQLMetrics.scala]: It seems `LongSQLMetricParam` is the simplest one. Don't have another good name in mind at the moment.

**Comment by carsonwang** [comment]
Thanks @vanzin and @chenghao-intel for reviewing. Just pushed updates to address the comments.

File: core/src/main/scala/org/apache/spark/util/JsonProtocol.scala [hunk 21]
```diff
@@ -19,19 +19,22 @@ package org.apache.spark.util
...
 import java.util.{Properties, UUID}
 
-import org.apache.spark.scheduler.cluster.ExecutorInfo
-
 import scala.collection.JavaConverters._
 import scala.collection.Map
 
+import com.fasterxml.jackson.databind.ObjectMapper
+import com.fasterxml.jackson.module.scala.DefaultScalaModule
+
```
Comment (vanzin) [inline core/src/main/scala/org/apache/spark/util/JsonProtocol.scala]: super nit: get rid of this empty line.

**Comment by vanzin** [comment]
LGTM. retest this please

**Comment by carsonwang** [comment]
retest this please

**Comment by vanzin** [comment]
retest this please

**Comment by steveloughran** [comment]
While I support this feature, people already see problems with scanning/loading app histories off the filesystem. This has the potential to increase it significantly. I think it helps justify having those fs logs save some metadata summary (started, ended, spark version,  ...) alongside the logs, so at least the app listing operation can bypass the scan. Separate JIRA, obviously

**Comment by vanzin** [comment]
@steveloughran while that's true, this change shouldn't make things that much worse; things are already pretty bad for applications that generate large logs.

I'm merging this to master.

**Comment by JoshRosen** [comment]
I just observed a SQLListenerMemoryLeakSuite memory leak test failure in Jenkins. Do you think that it could somehow be related to this patch? 

Here's the failing test:

https://amplab.cs.berkeley.edu/jenkins/job/Spark-Master-Maven-pre-YARN/HADOOP_VERSION=2.0.0-mr1-cdh4.1.2,label=spark-test/4896/

No need to revert or take drastic action yet, but let's keep an eye on that suite to see if the test fails in any more builds; you can monitor this test at https://spark-tests.appspot.com/tests/org.apache.spark.sql.execution.ui.SQLListenerMemoryLeakSuite/no%20memory%20leak, an experimental dashboard that I've been building to help investigate these types of flaky test issues.

**Comment by vanzin** [comment]
I can't really say, I'm not too familiar with the SQL backend here. @chenghao-intel took a look, maybe he has some insight.

**Comment by chenghao-intel** [comment]
Seems failed again, we are debugging on it now, will keep you posted if anything find.

**Comment by carsonwang** [comment]
I just submitted #9991 to fix the test failure. Details are described in the new PR. Thanks all!

**Comment by JoshRosen** [comment]
Just curious: are there any end-to-end tests for this feature? Would it be very hard to add some?

File: sql/core/src/main/scala/org/apache/spark/sql/SQLContext.scala [hunk 22]
```diff
@@ -1332,9 +1337,13 @@ object SQLContext {
    * Create a SQLListener then add it into SparkContext, and create an SQLTab if there is SparkUI.
    */
   private[sql] def createListenerAndUI(sc: SparkContext): SQLListener = {
-    val listener = new SQLListener(sc.conf)
-    sc.addSparkListener(listener)
-    sc.ui.foreach(new SQLTab(listener, _))
-    listener
+    if (sqlListener.get() == null) {
```
Comment (JoshRosen) [inline sql/core/src/main/scala/org/apache/spark/sql/SQLContext.scala]: Given that it's still legal to have multiple active SQLContexts in an application (although this is not a good practice), isn't this a user-facing behavioral change? If a user creates multiple SQLContexts, then they will now only have one SQL tab in the UI and one SQLListener, which is different than the old behavior.

**Comment by vanzin** [comment]
I don't think so. Tests would have to be added to one of the `sql/` modules since you can't really test those events from the `core/` module...

In `core/` the only thing that can be tested is that Jackson is doing the right thing when unknown event types are written or present in the event log.

File: sql/core/src/main/scala/org/apache/spark/sql/SQLContext.scala [hunk 23]
```diff
@@ -1332,9 +1337,13 @@ object SQLContext {
    * Create a SQLListener then add it into SparkContext, and create an SQLTab if there is SparkUI.
```
Comment (JoshRosen) [inline sql/core/src/main/scala/org/apache/spark/sql/SQLContext.scala]: If this method is going to be a no-op under certain circumstances then those circumstances should be documented here.

File: sql/core/src/main/scala/org/apache/spark/sql/SQLContext.scala (hunk 22 above)
Comment (JoshRosen) [inline sql/core/src/main/scala/org/apache/spark/sql/SQLContext.scala]: Also, is the only purpose of `sqlListener` to prevent multiple listeners from being created at the same time? If so, I think it would be better to use an AtomicBoolean so that we don't create another strong reference to a SQLListener, which might have a lot of internal state that could lead to memory leaks.

**Comment by JoshRosen** [comment]
@vanzin, in that case, why can't those tests live in the SQL module? I think that we should be able to add a test where we create a SparkContext + SQLContext with event logging enabled, run some queries, stop the context, then programmatically instantiate a HistoryServer, reconstitute the UI from the event logs, and perform some assertions over the loaded UI (e.g. check the number of SQL tabs, check that clicking on the SQL tab returns a page with some expected strings, etc). We could easily define some test helper classes to make this easy to write.

**Comment by vanzin** [comment]
I'm not saying they can't live there, I'm just saying they have to be written (since they don't exist) and when they are, they'll have to live there.

**Comment by JoshRosen** [comment]
Because this PR broke the Master Maven tests, #9991 isn't read to merge, and there's not an easy way to ignore failing tests (since the problem is broader in scope than that single test), I'm going to revert this for now. Let's aim to resubmit a new PR which addresses the concerns being discussed in #9991.
